usaspending_api/download/helpers/elasticsearch_download_functions.py

```diff
@@ -0,0 +1,95 @@
+import itertools
+import logging
+from abc import ABCMeta, abstractmethod
+from typing import Union
+
+from django.conf import settings
+from django.db.models import QuerySet
+from elasticsearch_dsl import A
+
+from usaspending_api.common.elasticsearch.search_wrappers import AwardSearch, TransactionSearch
+from usaspending_api.common.query_with_filters import QueryWithFilters
+from usaspending_api.search.models import AwardSearchView, UniversalTransactionView
+
+logger = logging.getLogger(__name__)
+
+
+class _ElasticsearchDownload(metaclass=ABCMeta):
+    _source_field = None
+    _filter_query_func = None
+    _search_type = None
+
+    @classmethod
+    def _get_download_ids_generator(cls, search: Union[AwardSearch, TransactionSearch], size: int):
+        """
+        Takes an AwardSearch or TransactionSearch object (that specifies the index, filter, and source) and returns
+        a generator that yields list of IDs in chunksize SIZE.
+        """
+        max_retries = 10
+        total = search.handle_count(retries=max_retries)
+        if total is None:
+            logger.error("Error retrieving total results. Max number of attempts reached.")
+            return
+        max_iterations = settings.MAX_DOWNLOAD_LIMIT // size
+        req_iterations = (total // size) + 1
+        num_iterations = min(max(1, req_iterations), max_iterations)
+        for iteration in range(num_iterations):
+            aggregation = A(
+                "terms",
+                field=cls._source_field,
+                include={"partition": iteration, "num_partitions": num_iterations},
+                size=size,
+            )
+            search.aggs.bucket("results", aggregation)
+            response = search.handle_execute(retries=max_retries).to_dict()
+
+            if response is None:
+                raise Exception("Breaking generator, unable to reach cluster")
+            results = []
+            for bucket in response["aggregations"]["results"]["buckets"]:
+                results.append(bucket["key"])
+
+            yield results
+
+    @classmethod
+    def _get_download_ids(cls, filters: dict, size: int = 100000) -> QuerySet:
+        """
+        Takes a dictionary of the different download filters and returns a flattened list of ids.
+        """
+        filter_query = cls._filter_query_func(filters)
+        search = cls._search_type().filter(filter_query).source([cls._source_field])
+        ids = cls._get_download_ids_generator(search, size)
+        flat_ids = list(itertools.chain.from_iterable(ids))
+        logger.info(f"Found {len(flat_ids)} {cls._source_field} based on filters")
+        return flat_ids
+
+    @classmethod
+    @abstractmethod
+    def query(cls, filters: dict) -> QuerySet:
+        pass
+
+
+class AwardsElasticsearchDownload(_ElasticsearchDownload):
+    _source_field = "award_id"
```

Comment:
I know we switched to using `generated_unique_award_id` to identify awards on the website, do we want to use those here as well? I don't think it'll actually make any difference, but it would be more consistent.

Reply:
Was torn on this. I am not against changing it, but feel more comfortable using `award_id` in the API since it is the PK for the `vw_award_search` view.